mailspring: init at 1.7.8

toschmidt opened the pull request:
<!-- Nixpkgs has a lot of new incoming Pull Requests, but not enough people to review this constant stream. Even if you aren't a committer, we would appreciate reviews of other PRs, especially simple ones like package updates. Just testing the relevant package/service and leaving a comment saying what you tested, how you tested it and whether it worked would be great. List of open PRs: <https://github.com/NixOS/nixpkgs/pulls>, for more about reviewing contributions: <https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#chap-reviewing-contributions>. Reviewing isn't mandatory, but it would help out a lot and reduce the average time-to-merge for all of us. Thanks a lot if you do! -->
###### Motivation for this change

Packaged the mail client [Mailspring](https://github.com/Foundry376/Mailspring) at version 1.7.5

> Mailspring is a new version of Nylas Mail maintained by one of the original authors. It's faster, leaner, and shipping today!

Currently two pull requests for mailspring at version 1.6.3 exist: #69384 and #69027 

Syncing mails does not work in #69384 and #69027 uses buildFHSUserEnv although only one file needs patching.

This PR is based on the work done by rummik in #69384 and solves both problems for the latest version of mailspring.

The application needs `gnome-keyring` for storing passwords.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers
cc @

toschmidt committed:
maintainers: add toschmidt

toschmidt committed:
mailspring: init at 1.7.8

drewrisinger commented:
Closes #69027 #69384
@GrahamcOfBorg build mailspring

drewrisinger commented on pkgs/applications/networking/mailreaders/mailspring/default.nix:
```diff
@@ -0,0 +1,85 @@
+  ];
+
+  unpackPhase = /* sh */ ''
+    dpkg -x $src .
```
I'm honestly not too familiar with policies on using dpkg (i.e. pre-compiled binaries) in Nix. I know general preference is to build from source if at all possible, but it seems this ISN'T possible due to a private C++ section of ``mailspring``? Is that correct? Did you try building from the GitHub at all?

toschmidt replied:
@drewrisinger thanks for your review.

Yes, mailspring depends on a private repo, that implements syncing mails. I found no other way to get the binary than depackaging one of the provided builds.

drewrisinger commented on pkgs/applications/networking/mailreaders/mailspring/default.nix:
```diff
@@ -0,0 +1,85 @@
+{ stdenv
+, fetchurl
+, autoPatchelfHook
+, dpkg
+, glib
+, nss
+, gtk3
+, xorg
+, alsaLib
+, db
+, libkrb5
+, libsecret
+, udev
+, coreutils
+, openssl
```
Sort these alphabetically

toschmidt replied:
Ok

drewrisinger commented on pkgs/applications/networking/mailreaders/mailspring/default.nix:
```diff
@@ -0,0 +1,85 @@
+    autoPatchelfHook
+  ];
+
+  buildInputs = [
+    glib
+    nss
+    gtk3
+    xorg.libXtst
+    xorg.libXScrnSaver
+    xorg.libxkbfile
+    alsaLib
+    db
+    libkrb5
+    libsecret
+  ];
```
Sort these & other package lists below alphabetically

toschmidt replied:
Ok

drewrisinger commented on pkgs/applications/networking/mailreaders/mailspring/default.nix:
```diff
@@ -0,0 +1,85 @@
+    cp -ar ./usr/share $out
+
+    substituteInPlace $out/share/mailspring/resources/app.asar.unpacked/mailsync \
+      --replace realpath ${coreutils.out}/bin/realpath
+    substituteInPlace $out/share/mailspring/resources/app.asar.unpacked/mailsync \
+      --replace dirname ${coreutils.out}/bin/dirname
```
```suggestion
      --replace realpath ${coreutils}/bin/realpath \
      --replace dirname ${coreutils}/bin/dirname
```

drewrisinger commented on pkgs/applications/networking/mailreaders/mailspring/default.nix:
```diff
@@ -0,0 +1,85 @@
+      --replace dirname ${coreutils.out}/bin/dirname
+
+    ln -s $out/share/mailspring/mailspring $out/bin/mailspring
+    ln -s ${openssl.out}/lib/libcrypto.so $out/lib/libcrypto.so.1.0.0
```
```suggestion
    ln -s ${openssl}/lib/libcrypto.so $out/lib/libcrypto.so.1.0.0
```

toschmidt replied:
without `${openssl.out}` build fails

drewrisinger commented on pkgs/applications/networking/mailreaders/mailspring/default.nix:
```diff
@@ -0,0 +1,85 @@
+    license = licenses.gpl3;
+    maintainers = with maintainers; [ toschmidt ];
+    homepage = https://getmailspring.com;
+    platforms = [ "x86_64-linux" ];
```
```suggestion
    platforms = platforms.x86_64;
```

drewrisinger commented on pkgs/applications/networking/mailreaders/mailspring/default.nix:
```diff
@@ -0,0 +1,85 @@
+    '';
+    license = licenses.gpl3;
+    maintainers = with maintainers; [ toschmidt ];
+    homepage = https://getmailspring.com;
```
```suggestion
    homepage = "https://getmailspring.com";
    downloadPage = "https://github.com/Foundry376/Mailspring";
```
Quote per RFC 39? RFC49?

drewrisinger commented on pkgs/top-level/all-packages.nix:
```diff
@@ -26444,4 +26444,6 @@ in
 
   jitsi-meet-electron = callPackage ../applications/networking/instant-messengers/jitsi-meet-electron { };
 
+  mailspring = callPackage ../applications/networking/mailreaders/mailspring {};
```
This should be inserted alphabetically-ish. I would probably place it near thunderbird, there might be other mail clients in there? ``all-packages`` is loosely sorted by category then name.

toschmidt replied:
I placed it near thunderbird and mutt.

drewrisinger reviewed (changes requested):
Thanks for submitting your first PR! This looks like a well-researched PR. I'm honestly not terribly familiar with the Linux binary side of nixpkgs, but I know the general preference is to build packages from scratch in Nixpkgs if possible.

There's some formatting/best practices included below.
Also, when trying to review this, I ran into a merge conflict with the maintainers list (probably b/c you submitted a different PR & now there's a conflict). So please remove that commit from this with ``git rebase -i master``. Reviewers can't easily test building this until that's fixed.

toschmidt commented:
@drewrisinger  thanks for your review.

> Also, when trying to review this, I ran into a merge conflict with the maintainers list (probably b/c you submitted a different PR & now there's a conflict). So please remove that commit from this with `git rebase -i master`. Reviewers can't easily test building this until that's fixed.

I removed myself from the maintainer list (it's my only PR so there should not be any conflict in that file). Before merging, we should probably revert that unless someone else wants to maintain the package. I also updated mailspring to the latest version (1.7.6) and applied your suggestions.

drewrisinger commented:
Oh, sorry. If this is actually your first package, then a commit adding yourself as maintainer IS appropriate. Just make sure that it merges with ``master`` branch. I recommend rebasing, dropping the revert commit, and then ``git push -f`` that fixed branch. You should resolve whatever merge commit the maintainer list pops up.

Builds locally with ``nixpkgs-review pr 85492``
```
https://github.com/NixOS/nixpkgs/pull/85492
1 package built:
mailspring
```

drewrisinger commented:
Note: commits will also need squashed down to just the maintainer commit & ``mailspring: init at 1.7.6`` (PR title should also be changed) before this PR is accepted/merged.

toschmidt commented:
Ok, the PR should have no conflicts with master and I squashed all commits down.

drewrisinger commented on pkgs/applications/networking/mailreaders/mailspring/default.nix:
```diff
@@ -0,0 +1,84 @@
+    ln -s ${openssl.out}/lib/libcrypto.so $out/lib/libcrypto.so.1.0.0
+  '';
+
+  postFixup = /* sh */ ''
```
why the comment here?

drewrisinger reviewed (commented):
Few small notes still to be resolved by someone who knows more about packaging ``.deb`` better than I.

Builds locally, not tested.
``nixpkgs-review pr 85492``
```
https://github.com/NixOS/nixpkgs/pull/85492
1 package built:
mailspring
```

drewrisinger commented:
Closes #69027

drewrisinger commented:
Closes #69384

tyrion commented:
I tried upgrading this to the latest Mailspring release and it seems to work fine so far.
You just need to change the version and sha checksum:

```
  version = "1.7.8";

  src = fetchurl {
    url = "https://github.com/Foundry376/Mailspring/releases/download/${version}/mailspring-${version}-amd64.deb";
    sha256 = "207fbf813b6da018a5b848e5dc1194b5996daab39adbd873b2cecb0565c105ce";
  };
```

tyrion commented on pkgs/applications/networking/mailreaders/mailspring/default.nix:
```diff
@@ -0,0 +1,84 @@
+      Mailspring is an open-source mail client forked from Nylas Mail and built with Electron.
+      Mailspring's sync engine runs locally, but its source is not open.
+    '';
+    license = licenses.gpl3;
```
Mailspring is unfortunately not free software. This should be changed to "unfree".

drewrisinger reviewed (approved):
* [x] Diff LGTM
* [x] Builds Locally
* [x] After build, GUI launches
* [x] Still not sure about best method when using dpkg, but seems to build fine.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/181

tobkratz reviewed (approved):
I can get it work. 
LGTM

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-already-reviewed/2617/172

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-already-reviewed/2617/181

danieldk reviewed (approved):
Thanks for submitting this and sorry for having to wait so long for getting this merged. It seems that mailspring relies on wget for crash dumps:

```
$ wrapit --bind /tmp/.X11-unix/X0 /tmp/.X11-unix/X0 --dir /home/daniel --setenv DISPLAY :0 `realpath results/mailspring/bin/mailspring`
Fontconfig error: Cannot load default config file
Running database migrations
App load time: 145ms

Running Setup
{"error":null}
Cannot upload crash dump: cannot exec /usr/bin/wget
Failed to get crash dump id.
Report Id: 
Cannot upload crash dump: cannot exec /usr/bin/wget
Failed to get crash dump id.
Report Id: 
```

Could be something to fix for a future PR.
